lib/mobility/backends/active_record/container.rb

```diff
@@ -107,32 +108,22 @@ def initialize_dup(source)
       private
 
       def model_translations(locale)
-        model[column_name][locale] ||= {}
-      end
+        translations = model[column_name]
 
-      def set_attribute_translation(locale, value)
-        translations = model[column_name] || {}
-        translations[locale.to_s] ||= {}
-        translations[locale.to_s][attribute] = value
-        model[column_name] = translations
-      end
+        return unless translations
 
-      class Coder
-        def self.dump(obj)
-          if obj.is_a? ::Hash
-            obj.inject({}) do |translations, (locale, value)|
-              value.each do |k, v|
-                (translations[locale] ||= {})[k] = v if v.present?
-              end
-              translations
-            end
-          else
-            raise ArgumentError, "Attribute is supposed to be a Hash, but was a #{obj.class}. -- #{obj.inspect}"
-          end
-        end
+        translations[locale.to_s]
+      end
 
-        def self.load(obj)
-          obj
+      def set_attribute_translation(locale, value)
+        locale_translations = model_translations(locale)
+
+        if locale_translations
+          locale_translations[attribute.to_s] = value
+          locale_translations.compact!
+          model[column_name].compact!
```

Comment:
not sure if those two compacts impact performance or could be optimized

Reply:
I think it's enough to delete the key if `value` is not `present?`, and otherwise set it. `compact!` assumes we want to clean up other blank translations but that shouldn't be necessary.

so:

```ruby
if locale_translations
  if value.nil?
    locale_translations.delete(attribute.to_s)
  else
    locale_translations[attribute.to_s] = value
  end
end
```

None of these edge cases, mind you, is tested, much or at all. It might make sense to add a container spec which checks that when you set a value of `nil` to an attribute with the container backend, and reload, it comes back `nil`, _and_ the stored `translations` hash has no key for that locale/attribute.

The handling of `""` in the current implementation is wrong. `""` should be treated as a value; if someone wants blank strings treated as `nil` then they can use the presence plugin to do that.

Reply:
Doesn't `.compact` only clear `nil` values, so blank strings would not be affected by it and saved correctly?

```rb
{ 'a' => nil, 'b' => "" }.compact!
# => {"b"=>""}
```

Though if we assume the translation column is well formed (has no dangling empty hashes), it could indeed be done like you suggested without `compact` but with one addition to remove empty locale hashes (so we don't save `{ "en" => {} }` but `{}` after deleting the last translation):

```rb
if locale_translations
  if value.nil?
    locale_translations.delete(attribute.to_s)

    # delete empty locale hashes if we removed the last translation
    model[column_name].delete(locale.to_s) if locale_translations.empty?
  else
    locale_translations[attribute.to_s] = value
  end
else 
  ...
end
```

I'll update the implementation and add some specs like you requested (so you can play around with it more easily, too).